### PR TITLE
Make locking 'idempotent' in consul coordination (closes #61)

### LINKED ISSUE #61: Make locking 'idempotent' in consul coordination
After fixing #59, a review of locking stage in consul coordination is needed. See if it's neccessary to update the full TTL after a retry.

### PR DESCRIPTION
**Pull request by juanjovazquez**

### CODE REVIEW

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala [hunk 1]
```diff
@@ -63,13 +64,46 @@ final class ConsulCoordination(prefix: String, clusterName: String, host: String
...
 
   override def lock[N](self: N, ttl: Duration)(implicit ec: ExecutionContext, mat: Materializer): Future[LockResult] = {
-    // TODO Make idempotent wrt retries! Actually a retry should not be fully idempotent, but update to the full TTL.
     val uriLock = baseUri.withPath(baseUri.path / "lock")
-    val response = for {
-      sessionId <- createSession(ttl)
-      result <- putKeyWithSession(uriLock, sessionId)
-    } yield result
-    response.map(isLocked => if (isLocked) LockResult.Success else LockResult.Failure)
+    val body = HttpEntity(`application/json`, self.toString)
```
Comment (juanjovazquez) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Consul API expects values in the message body

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala [hunk 2]
```diff
@@ -63,13 +64,46 @@ final class ConsulCoordination(prefix: String, clusterName: String, host: String
...
+    def readLock() = {
+      def unmarshallLockHolder(entity: ResponseEntity) = {
+        def toLockHolder(s: String) = {
+          import rapture.json._
+          import rapture.json.jsonBackends.spray._
+          def jsonToNode(json: Json) = {
+            val value = json.Value.as[String]
+            new String(decode(value))
```
Comment (juanjovazquez) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Returned values are base64 encoded.

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala [hunk 3]
```diff
@@ -95,9 +129,9 @@ final class ConsulCoordination(prefix: String, clusterName: String, host: String
 
   override def initialBackendContext = ""
 
-  private def putKeyWithSession(keyUri: Uri, sessionId: SessionId)(implicit ec: ExecutionContext, mat: Materializer): Future[Boolean] = {
+  private def putKeyWithSession(keyUri: Uri, content: Strict, sessionId: SessionId)(implicit ec: ExecutionContext, mat: Materializer): Future[Boolean] = {
```
Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Two things:
1. Code style: if you have to use `Strict`, please use it qualified `HttpEntity.Strict` (like `HttpEntity.Empty` above) and remove the respective import above
2. Is it really necessary to use a strict entity? Please try `HttpEntity` or `RequetEntity`.

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala (hunk 2 above)
Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Please use `new String(..., UTF_8)` (see `de/heikoseeberger/constructr/akka/package.scala`).

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala [hunk 4]
```diff
@@ -63,13 +64,46 @@ final class ConsulCoordination(prefix: String, clusterName: String, host: String
...
-    response.map(isLocked => if (isLocked) LockResult.Success else LockResult.Failure)
+    val body = HttpEntity(`application/json`, self.toString)
+    def readLock() = {
+      def unmarshallLockHolder(entity: ResponseEntity) = {
+        def toLockHolder(s: String) = {
+          import rapture.json._
+          import rapture.json.jsonBackends.spray._
+          def jsonToNode(json: Json) = {
+            val value = json.Value.as[String]
+            new String(decode(value), UTF_8)
```
Comment (juanjovazquez) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Decoding with UTF8 charset.

**Comment by juanjovazquez** [comment]
@hseeberger please, check if this fits you. thanks!.

**Comment by hseeberger** [comment]
Looks great. Thanks!

**Pull request merged**